File: odk/examples/DevelopersGuide/Charts/ChartInCalc.java
/* -*- Mode: Java; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*************************************************************************
 *
 *  The Contents of this file are made available subject to the terms of
 *  the BSD license.
 *
 *  Copyright 2000, 2010 Oracle and/or its affiliates.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. Neither the name of Sun Microsystems, Inc. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 *  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 *  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *************************************************************************/

// __________ Imports __________

import com.sun.star.uno.UnoRuntime;
import com.sun.star.lang.*;

// property access
import com.sun.star.beans.*;

// application specific classes
import com.sun.star.chart.*;
import com.sun.star.drawing.*;

import com.sun.star.table.CellRangeAddress;
import com.sun.star.table.XCellRange;
import com.sun.star.sheet.XCellRangeAddressable;

import com.sun.star.frame.XModel;
import com.sun.star.util.XNumberFormatsSupplier;
import com.sun.star.util.XNumberFormats;

// base graphics things
import com.sun.star.awt.Point;
import com.sun.star.awt.Size;
import com.sun.star.awt.FontWeight;
// Exceptions
import com.sun.star.uno.Exception;
import com.sun.star.uno.RuntimeException;
import com.sun.star.beans.UnknownPropertyException;
import com.sun.star.lang.IndexOutOfBoundsException;
import com.sun.star.util.MalformedNumberFormatException;


// __________ Implementation __________

// Create a spreadsheet add some data and add a chart

public class ChartInCalc
{


    public static void main( String args[] )
    {
        Helper aHelper = new Helper( args );

        CalcHelper aCalcHelper = new CalcHelper( aHelper.createSpreadsheetDocument() );

        // insert a cell range with 4 columns and 24 rows filled with random numbers
        XCellRange aRange = aCalcHelper.insertRandomRange( 4, 24 );
        CellRangeAddress aRangeAddress = UnoRuntime.queryInterface(
            XCellRangeAddressable.class, aRange).getRangeAddress();

        // change view to sheet containing the chart
        aCalcHelper.raiseChartSheet();

        // the unit for measures is 1/100th of a millimeter
        // position at (1cm, 1cm)
        Point aPos    = new Point( 1000, 1000 );

        // size of the chart is 15cm x 9.271cm
        Size  aExtent = new Size( 15000, 9271 );

        // insert a new chart into the "Chart" sheet of the
        // spreadsheet document
        XChartDocument aChartDoc = aCalcHelper.insertChart(
            "ScatterChart",
            aRangeAddress,
            aPos,
            aExtent,
            "com.sun.star.chart.XYDiagram" );

        // instantiate test class with newly created chart
        ChartInCalc aTest   = new ChartInCalc( aChartDoc );

        try
        {
            aTest.lockControllers();

            aTest.testDiagram();
            aTest.testArea();
            aTest.testWall();
            aTest.testTitle();
            aTest.testAxis();
            aTest.testGrid();

            // show an intermediate state, ...
            aTest.unlockControllers();
            aTest.lockControllers();

            // ..., because the following takes a while:
            // an internet URL has to be resolved
            aTest.testDataRowProperties();
            aTest.testDataPointProperties();

            aTest.unlockControllers();
        }
        catch( Exception ex )
        {
            System.out.println( "UNO Exception caught: " + ex );
            System.out.println( "Message: " + ex.getMessage() );
        }

        System.exit( 0 );
    }




    public ChartInCalc( XChartDocument aChartDoc )
    {
        maChartDocument = aChartDoc;
        maDiagram       = maChartDocument.getDiagram();
    }



    public void lockControllers()
        throws RuntimeException
    {
        UnoRuntime.queryInterface( XModel.class, maChartDocument ).lockControllers();
    }



    public void unlockControllers()
        throws RuntimeException
    {
        UnoRuntime.queryInterface( XModel.class, maChartDocument ).unlockControllers();
    }



    public void testDiagram()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet aDiaProp = UnoRuntime.queryInterface( XPropertySet.class, maDiagram );

        if( aDiaProp != null )
        {
            // change chart type
            aDiaProp.setPropertyValue( "Lines", Boolean.TRUE);

            // change attributes for all series
            // set line width to 0.5mm
            aDiaProp.setPropertyValue( "LineWidth",  Integer.valueOf( 50 ));
        }
    }



    public void testDataRowProperties()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        // change properties of the data series
        try
        {
            XPropertySet aSeriesProp;
            for( int i = 1; i <= 3; i++ )
            {
                aSeriesProp = maDiagram.getDataRowProperties( i );
                aSeriesProp.setPropertyValue( "LineColor", Integer.valueOf(
                                                  0x400000 * i +
                                                  0x005000 * i +
                                                  0x0000ff - 0x40 * i ));
                if( 1 == i )
                {
                    StringBuffer sUrl = new StringBuffer("file:///");
                    try {
                        /* for use without net it's easier to load a local graphic */
                        java.io.File sourceFile = new java.io.File("bullet.gif");
                        sUrl.append(sourceFile.getCanonicalPath().replace('\\', '/'));
                    } catch (java.io.IOException e) {
                        sUrl = new StringBuffer("http://graphics.openoffice.org/chart/bullet1.gif");
                    }

                    // set a bitmap via URL as symbol for the first series
                    aSeriesProp.setPropertyValue( "SymbolType", Integer.valueOf( ChartSymbolType.BITMAPURL ));
                    aSeriesProp.setPropertyValue( "SymbolBitmapURL", sUrl.toString() );
                }
                else
                {
                    aSeriesProp.setPropertyValue( "SymbolType", Integer.valueOf( ChartSymbolType.SYMBOL1 ));
                    aSeriesProp.setPropertyValue( "SymbolSize", new Size( 250, 250 ));
                }
            }
        }
        catch( IndexOutOfBoundsException ex )
        {
            System.out.println( "Oops, there not enough series for setting properties: " + ex );
        }
    }



    public void testDataPointProperties()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        // set properties for a single data point
        try
        {
            // determine the maximum value of the first series

            XChartDataArray aDataArray = UnoRuntime.queryInterface(
                XChartDataArray.class, maChartDocument.getData());
            double aData[][] = aDataArray.getData();

            int i;
            double fMax = aData[ 0 ][ 1 ];
            for( i = 1; i < aData.length; i++ )
            {
                if( aData[ i ][ 1 ] > fMax )
                {
                    fMax = aData[ i ][ 1 ];
                }
            }

            // first parameter is the index of the point, the second one is the series
            XPropertySet aPointProp = maDiagram.getDataPointProperties( 0, 1 );

            // set a different, larger symbol
            aPointProp.setPropertyValue( "SymbolType", Integer.valueOf( ChartSymbolType.SYMBOL6 ));
            aPointProp.setPropertyValue( "SymbolSize", new Size( 600, 600 ));

            // add a label text with bold font, bordeaux red 14pt
            aPointProp.setPropertyValue( "DataCaption", Integer.valueOf( ChartDataCaption.VALUE ));
            aPointProp.setPropertyValue( "CharHeight",  new Float( 14.0 ));
            aPointProp.setPropertyValue( "CharColor",   Integer.valueOf( 0x993366 ));
            aPointProp.setPropertyValue( "CharWeight",  new Float( FontWeight.BOLD ));
        }
        catch( IndexOutOfBoundsException ex )
        {
            System.out.println( "Oops, there not enough data points or series for setting properties: " + ex );
        }
    }



    public void testArea()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet   aArea = maChartDocument.getArea();

        if( aArea != null )
        {
            // change background color of entire chart
            aArea.setPropertyValue( "FillStyle", FillStyle.SOLID );
            aArea.setPropertyValue( "FillColor", Integer.valueOf( 0xeeeeee ));
        }
    }



    public void testWall()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet aWall = UnoRuntime.queryInterface(
                                  X3DDisplay.class, maDiagram ).getWall();

        // change background color of area
        aWall.setPropertyValue( "FillStyle", FillStyle.SOLID );
        aWall.setPropertyValue( "FillColor", Integer.valueOf( 0xcccccc ));
    }



    public void testTitle()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        // change main title
        XPropertySet aDocProp = UnoRuntime.queryInterface(
            XPropertySet.class, maChartDocument );
        aDocProp.setPropertyValue( "HasMainTitle", Boolean.TRUE);

        XShape aTitle = maChartDocument.getTitle();
        XPropertySet aTitleProp = UnoRuntime.queryInterface( XPropertySet.class, aTitle );

        // set new text
        if( aTitleProp != null )
        {
            aTitleProp.setPropertyValue( "String", "Random Scatter Chart" );
            aTitleProp.setPropertyValue( "CharHeight", new Float(14.0) );
        }

        // align title with y axis
        XShape aAxis = UnoRuntime.queryInterface(
            XShape.class, UnoRuntime.queryInterface(
                XAxisYSupplier.class, maDiagram ).getYAxis() );

        if( aAxis != null &&
            aTitle != null )
        {
            Point aPos = aTitle.getPosition();
            aPos.X = ( aAxis.getPosition() ).X;
            aTitle.setPosition( aPos );
        }
    }



    public void testAxis()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException,
               MalformedNumberFormatException
    {
        // x axis
        XPropertySet aAxisProp = UnoRuntime.queryInterface(
                                      XAxisXSupplier.class, maDiagram ).getXAxis();
        if( aAxisProp != null )
        {
            aAxisProp.setPropertyValue( "Max",      Integer.valueOf( 24 ));
            aAxisProp.setPropertyValue( "StepMain", Integer.valueOf( 3 ));
        }

        // change number format for y axis
        aAxisProp = UnoRuntime.queryInterface(
                         XAxisYSupplier.class, maDiagram ).getYAxis();

        // add a new custom number format and get the new key
        int nNewNumberFormat = 0;
        XNumberFormatsSupplier aNumFmtSupp = UnoRuntime.queryInterface(
            XNumberFormatsSupplier.class, maChartDocument );

        if( aNumFmtSupp != null )
        {
            XNumberFormats aFormats = aNumFmtSupp.getNumberFormats();
            Locale aLocale = new Locale( "de", "DE", "de" );

            String aFormatStr = aFormats.generateFormat( nNewNumberFormat, aLocale, true, true, (short)3, (short)1 );
            nNewNumberFormat = aFormats.addNew( aFormatStr, aLocale );
        }

        if( aAxisProp != null )
        {
            aAxisProp.setPropertyValue( "NumberFormat", Integer.valueOf( nNewNumberFormat ));
        }
    }



    public void testGrid()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        // y major grid
        XPropertySet aGridProp = UnoRuntime.queryInterface(
            XPropertySet.class,
            UnoRuntime.queryInterface(
                XAxisYSupplier.class, maDiagram ).getYMainGrid());

        if( aGridProp != null )
        {
            LineDash aDash = new LineDash();
            aDash.Style    = DashStyle.ROUND;
            aDash.Dots     = 2;
            aDash.DotLen   = 10;
            aDash.Dashes   = 1;
            aDash.DashLen  = 200;
            aDash.Distance = 100;

            aGridProp.setPropertyValue( "LineColor", Integer.valueOf( 0x999999 ));
            aGridProp.setPropertyValue( "LineStyle", LineStyle.DASH );
            aGridProp.setPropertyValue( "LineDash", aDash );
            aGridProp.setPropertyValue( "LineWidth", Integer.valueOf( 30 ));
        }
    }




    // private members


    private final XChartDocument maChartDocument;
    private final XDiagram       maDiagram;
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: odk/examples/DevelopersGuide/Charts/python/ChartInCalc.py
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import os

import Helper
import CalcHelper

from com.sun.star.awt import FontWeight, Point, Size
from com.sun.star.chart import ChartSymbolType, ChartDataCaption
from com.sun.star.drawing import LineDash
from com.sun.star.drawing.DashStyle import ROUND
from com.sun.star.drawing.FillStyle import SOLID
from com.sun.star.drawing.LineStyle import DASH
from com.sun.star.lang import IndexOutOfBoundsException, Locale


class ChartInCalc(object):
    def __init__(self, chart_doc):
        super(ChartInCalc, self).__init__()
        self._chart_document = chart_doc
        self._diagram = self._chart_document.getDiagram()

    def lock_controllers(self):
        self._chart_document.lockControllers()

    def unlock_controllers(self):
        self._chart_document.unlockControllers()

    def test_diagram(self):
        dia_prop = self._diagram

        if dia_prop is not None:
            # change chart type
            dia_prop.setPropertyValue("Lines", True)
            # change attributes for all series
            # set line width to 0.5mm
            dia_prop.setPropertyValue("LineWidth", 50)

    def test_data_row_properties(self):
        # change properties of the data series
        try:
            for i in range(4):
                series_prop = self._diagram.getDataRowProperties(i)
                series_prop.setPropertyValue("LineColor", int(0x400000 * i + 0x005000 * i + 0x0000FF - 0x40 * i))

                if i == 1:
                    source_file = "bullet.gif"
                    url = os.path.abspath(source_file).replace("\\", "/")

                    if os.path.exists(url):
                        url = "file:///" + url
                    else:
                        url = "http://graphics.openoffice.org/chart/bullet1.gif"

                    series_prop.setPropertyValue("SymbolType", int(ChartSymbolType.BITMAPURL))
                    series_prop.setPropertyValue("SymbolBitmapURL", url)
                else:
                    series_prop.setPropertyValue("SymbolType", int(ChartSymbolType.SYMBOL1))
                    series_prop.setPropertyValue("SymbolSize", Size(250, 250))

        except IndexOutOfBoundsException as err:
            print(f"Oops, there not enough series for setting properties: {err}")

    def test_data_point_properties(self):
        #  set properties for a single data point
        try:
            # first parameter is the index of the point, the second one is the series
            point_prop = self._diagram.getDataPointProperties(0, 1)

            # set a different, larger symbol
            point_prop.setPropertyValue("SymbolType", int(ChartSymbolType.SYMBOL6))
            point_prop.setPropertyValue("SymbolSize", Size(600, 600))

            # add a label text with bold font, bordeaux red 14pt
            point_prop.setPropertyValue("DataCaption", int(ChartDataCaption.VALUE))
            point_prop.setPropertyValue("CharHeight", 14.0)
            point_prop.setPropertyValue("CharColor", int(0x993366))
            point_prop.setPropertyValue("CharWeight", FontWeight.BOLD)

        except IndexOutOfBoundsException as err:
            print(f"Oops, there not enough series for setting properties: {err}")

    def test_area(self):
        area = self._chart_document.getArea()
        if area is not None:
            # change background color of entire chart
            area.setPropertyValue("FillStyle", SOLID)
            area.setPropertyValue("FillColor", int(0xEEEEEE))

    def test_wall(self):
        wall = self._diagram.getWall()

        # change background color of area
        wall.setPropertyValue("FillStyle", SOLID)
        wall.setPropertyValue("FillColor", int(0xCCCCCC))

    def test_title(self):
        # change main title
        doc_prop = self._chart_document
        doc_prop.setPropertyValue("HasMainTitle", True)

        title = self._chart_document.getTitle()

        # set new text
        if title is not None:
            title.setPropertyValue("String", "Random Scatter Chart")
            title.setPropertyValue("CharHeight", 14.0)

        # align title with y axis
        axis = self._diagram.getYAxis()

        if axis is not None and title is not None:
            pos = title.getPosition()
            pos.X = axis.getPosition().X
            title.setPosition(pos)

    def test_axis(self):
        # x axis
        axis = self._diagram.getXAxis()

        if axis is not None:
            axis.setPropertyValue("Max", 24)
            axis.setPropertyValue("StepMain", 3)

        # change number format for y axis
        axis = self._diagram.getYAxis()

        # add a new custom number format and get the new key
        new_number_format = 0
        num_fmt_supp = self._chart_document

        if num_fmt_supp is not None:
            formats = num_fmt_supp.getNumberFormats()
            locale = Locale("de", "DE", "de")

            format_str = formats.generateFormat(new_number_format, locale, True, True, 3, 1)
            new_number_format = formats.addNew(format_str, locale)

        if axis is not None:
            axis.setPropertyValue("NumberFormat", int(new_number_format))

    def test_grid(self):
        # y major grid
        grid = self._diagram.getYMainGrid()

        if grid is not None:
            dash = LineDash()
            dash.Style = ROUND
            dash.Dots = 2
            dash.DotLen = 10
            dash.Dashes = 1
            dash.DashLen = 200
            dash.Distance = 100

            grid.setPropertyValue("LineColor", int(0x999999))
            grid.setPropertyValue("LineStyle", DASH)
            grid.setPropertyValue("LineDash", dash)
            grid.setPropertyValue("LineWidth", 30)


def main():
    # Create a spreadsheet add some data and add a chart
    helper = Helper.Helper()

    calc_helper = CalcHelper.CalcHelper(helper.create_spreadsheet_document())

    # insert a cell range with 4 columns and 24 rows filled with random numbers
    cell_range = calc_helper.insert_random_range(4, 24)
    range_address = cell_range.RangeAddress

    # change view to sheet containing the chart
    calc_helper.raise_chart_sheet()

    # the unit for measures is 1/100th of a millimeter
    # position at (1cm, 1cm)
    pos = Point(1000, 1000)

    # size of the chart is 15cm x 9.271cm
    extent = Size(15000, 9271)

    # insert a new chart into the "Chart" sheet of the spreadsheet document
    chart_doc = calc_helper.insert_chart("ScatterChart", range_address, pos, extent, "com.sun.star.chart.XYDiagram")

    test = ChartInCalc(chart_doc)

    try:
        test.lock_controllers()

        test.test_diagram()
        test.test_area()
        test.test_wall()
        test.test_title()
        test.test_axis()
        test.test_grid()

        # show an intermediate state, ...
        test.unlock_controllers()
        test.lock_controllers()

        # ..., because the following takes a while:
        # an internet URL has to be resolved
        test.test_data_row_properties()
        test.test_data_point_properties()

        test.unlock_controllers()

    except Exception as err:
        print(f"UNO Exception caught: {err}")


# Main entry point
if __name__ == "__main__":
    main()
